src/uct/ib/base/ib_iface.c

```diff
@@ -548,8 +549,25 @@ ucs_status_t uct_ib_iface_get_device_address(uct_iface_h tl_iface,
     return UCS_OK;
 }
 
+static int uct_ib_iface_roce_is_reachable_via_routing(unsigned subnet_mask,
+        const uct_ib_device_gid_info_t *local_gid_info,
+        const uct_ib_address_t *remote_ib_addr)
+{
+    size_t size;
+
+    if (!uct_ib_device_is_potentially_reachable(subnet_mask, local_gid_info)) {
```

Comment:
seems this function should return 1 if it's RoCEv2 and compare addresses if it's ROCEv1

Reply:
are you sure? I seem to remember only RoCEv2 has IP addresses, and RoCEv1has IB GRH and nothing to apply the netmask to.

Reply:
so we don't really have a way to know in the code if something is reachable or not according to the netmask, right? so i don't understand the point, do you need to add option to filter by address/mask like in IB?

Reply:
In RoCEv1 - no, I can't think of a good way, so we can leave it "always reachable" (like it is today). In RoCEv2, which uses IPs - this code will determine reachability based on the netmask, similarly to OS routing.

Reply:
In ROCEv2 it could be reachable also if netmask does not match, in case there is a router in the network

Reply:
> In ROCEv2 it could be reachable also if netmask does not match, in case there is a router in the network

@yosefe - this is exactly what caused the problem in the first place: UCX assumes there is a router - but actually there isn't, and packets fail to reach the target (causing a fatal timeout). This patch adds support for networks WITHOUT a router, where the routing is done solely based on the IP address and subnet mask of each port. I'll put a boolean user configuration to enable (can assume router presence by default).

Reply:
ok, let's have a config which is used on rocev2 only which limits reachability to same subnet (as configured in the system)